src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,733 @@
+
+        if (pendingInEffect) {
+            delayInfo.delay = delayInfo.pendingDelay;
+        }
+
+        delayInfo.pendingDelay = delay;
+        delayInfo.pendingDelayEffectTimestamp = uint32(block.timestamp + ALLOCATION_DELAY_CONFIGURATION_DELAY);
+
+        emit AllocationDelaySet(operator, delay);
+    }
+
+    /**
+     * @notice For a single strategy, update freeMagnitude by adding completable pending free magnitudes
+     * @param operator address to update freeMagnitude for
+     * @param strategy the strategy to update freeMagnitude for
+     * @param numToComplete the number of pending free magnitudes deallocations to complete
+     * @dev read through pending free magnitudes and add to freeMagnitude if completableTimestamp is >= block timestamp
+     * In addition to updating freeMagnitude, updates next starting index to read from for pending free magnitudes after completing
+     */
+    function _updateFreeMagnitude(address operator, IStrategy strategy, uint16 numToComplete) internal {
+        OperatorMagnitudeInfo storage info = operatorMagnitudeInfo[operator][strategy];
+        (uint64 freeMagnitudeToAdd, uint192 completed) =
+            _getPendingFreeMagnitude(operator, strategy, numToComplete, info.nextPendingFreeMagnitudeIndex);
+        info.freeMagnitude += freeMagnitudeToAdd;
+        info.nextPendingFreeMagnitudeIndex += completed;
+    }
+
+    /**
+     * @notice For a single strategy, modify magnitude allocations for each of the specified operatorSets
+     * @param operator address to modify allocations for
+     * @param allocation the magnitude allocations to modify for a single strategy
+     * @param allocationEffectTimestamp the timestamp when the allocations will take effect
+     * @param deallocationCompletableTimestamp the timestamp when the deallocations will be completable
+     * @dev For each allocation, allocation.operatorSets MUST be ordered in ascending order according to the
+     * encoding of the operatorSet. This is to prevent duplicate operatorSets being passed in. The easiest way to ensure
+     * ordering is to sort allocated operatorSets by address first, and then sort for each avs by ascending operatorSetIds.
+     */
+    function _modifyAllocations(
+        address operator,
+        MagnitudeAllocation calldata allocation,
+        uint32 allocationEffectTimestamp,
+        uint32 deallocationCompletableTimestamp
+    ) internal {
+        require(allocation.operatorSets.length == allocation.magnitudes.length, InputArrayLengthMismatch());
+
+        // OperatorSet[] calldata opSets = allocation.operatorSets;
+
+        for (uint256 i = 0; i < allocation.operatorSets.length; ++i) {
+            require(
+                avsDirectory.isOperatorSet(allocation.operatorSets[i].avs, allocation.operatorSets[i].operatorSetId),
+                InvalidOperatorSet()
+            );
```

Comment:
it strikes me that this might be a lot of duplicate calls.
should we have an arrayified version of `isOperatorSet` (e.g. `areOperatorSets`)?

Reply:
`isOperatorSetBatch` has been added